src/main/java/com/powsybl/openloadflow/ac/nr/CustomNewtonRaphsonStoppingCriteria.java

```diff
@@ -0,0 +1,85 @@
+/**
+ * Copyright (c) 2019, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.powsybl.openloadflow.ac.nr;
+
+import com.powsybl.openloadflow.ac.equations.AcEquationType;
+import com.powsybl.openloadflow.ac.equations.AcVariableType;
+import com.powsybl.openloadflow.equations.EquationSystem;
+import com.powsybl.openloadflow.equations.Vectors;
+
+/**
+ * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
+ */
+public class CustomNewtonRaphsonStoppingCriteria implements NewtonRaphsonStoppingCriteria {
+
+    /**
+     * Convergence epsilon per equation: 10^-4 in p.u => 10^-2 in Kv, Mw or MVar
+     */
+    //FIXME what to do with this value?
+    public static final double DEFAULT_CONV_EPS_PER_EQ = Math.pow(10, -4);
+
+    //FIXME what to do with this value?
+    private final double convEpsPerEq;
+
+    private final double maxActivePowerMismatch;
+
+    private final double maxReactivePowerMismatch;
+
+    private final double maxVoltageMismatch;
+
+    public CustomNewtonRaphsonStoppingCriteria(double convEpsPerEq, double maxActivePowerMismatch,
+                                               double maxReactivePowerMismatch, double maxVoltageMismatch) {
+        this.convEpsPerEq = convEpsPerEq;
+        this.maxActivePowerMismatch = maxActivePowerMismatch;
+        this.maxReactivePowerMismatch = maxReactivePowerMismatch;
+        this.maxVoltageMismatch = maxVoltageMismatch;
+    }
+
+    private double computeNorm(double[] fx) {
+        return Vectors.norm2(fx);
+    }
+
+    private boolean computeStop(double[] fx, EquationSystem<AcVariableType, AcEquationType> equationSystem) {
+        for (var eq : equationSystem.getIndex().getSortedEquationsToSolve()) {
+            var type = eq.getType();
+            var idx = eq.getColumn();
+            switch (type) {
+                case BRANCH_TARGET_P:
+                case BUS_TARGET_P:
+                case DUMMY_TARGET_P:
+                    if (fx[idx] >= maxActivePowerMismatch) {
```

Comment:
For each test, you need to compute abs(fx[idx]) since fx contains the difference and not the absolute difference (thus can be negative).
Moreover, there is another difficult part here as the equations have been normalized in per unit, you need to recover the mismatch in physical units to compare to the criteria.
For power quantities this is easy because each equation is normalized by PerUnit.SB, see the PerUnit class.
However, for voltage quantities, they have been normalized according to "Nominal Voltage" at the corresponding bus. So you need to get the original bus whose equation is referring to.

Reply:
According to me, for volatge we want the threshold in pu